### DISCUSSION TITLE
Is there an example of a good logger implementation for socket.io?

### DISCUSSION THREAD

**Discussion by Natashkinsasha** [socketio/socket.io#4283] [Q&A]
I have been using socket.io for several years, but still have not found or come up with a way to log all messages, ack, emit, connections and disconnections. At the moment, the best thing I came up with is to log everything that is encoded and decoded in my custom parser. But unfortunately in the encoder and decoder there is no information about the socket on which the messages go. I am increasingly moving away from the REST API in favor of socket.io, but I am very frustrated that there is still no logging tool

**Comment by darrachequesne** [comment]
Hi! Do you have a specific API in mind?

> a way to log all messages, ack, emit, connections and disconnections

Currently, you should be able to log connections, disconnections and incoming packets:

```js
io.on("connection", (socket) => {
  // log connection

  socket.onAny((event, ...args) => {
    // log incoming packet
  });

  socket.on("disconnection", (reason) => {
    // log disconnection
  });
});
```

You could also intercept the output of the `debug` utility (by using the `DEBUG` env variable), but I guess it's too verbose for your use case.

**Reply by Natashkinsasha** [reply]
I would like a universal solution that is easy to connect. And what you suggest requires manual logging. It would be ideal if there would be a middleware for logging. I tried to write one, but failed.

```typescript
io.use(logger())
```
